File: app/src/main/java/tutorials/acviewmodel/home/SelectedRepoViewModel.java
package tutorials.acviewmodel.home;

import android.arch.lifecycle.LiveData;
import android.arch.lifecycle.MutableLiveData;
import android.arch.lifecycle.ViewModel;
import android.os.Bundle;
import android.util.Log;

import javax.inject.Inject;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import tutorials.acviewmodel.model.Repo;
import tutorials.acviewmodel.networking.RepoApi;
import tutorials.acviewmodel.networking.RepoService;

public class SelectedRepoViewModel extends ViewModel {

    private final MutableLiveData<Repo> selectedRepo = new MutableLiveData<>();
    private Call<Repo> repoCall;
    private final RepoService repoService;

    public LiveData<Repo> getSelectedRepo(){
        return selectedRepo;
    }

    @Inject
    public SelectedRepoViewModel(RepoService repoService){
        this.repoService = repoService;
    }

    void setSelectedRepo(Repo repo){
        selectedRepo.setValue(repo);
    }

    //<editor-fold desc="For saving and restoring data when android kills the process on out of memory exception or process death">
    public void restoreFromBundle(Bundle savedInstanceState) {
        if(selectedRepo.getValue() == null){
            //We only care about restoring if we don't have a selected Repo set already.
            if (savedInstanceState != null && savedInstanceState.containsKey("repo_details")) {
                loadRepo(savedInstanceState.getStringArray("repo_details"));
            }
        }
    }

    public void saveToBundle(Bundle outState) {
        if (selectedRepo.getValue() != null) {
            outState.putStringArray("repo_details", new String[]{
                    selectedRepo.getValue().owner().login(),
                    selectedRepo.getValue().name()
            });
        }
    }

    private void loadRepo(String[] repoDetails) {
        repoCall = repoService.getRepo(repoDetails[0], repoDetails[1]);
        repoCall.enqueue(new Callback<Repo>() {
            @Override
            public void onResponse(Call<Repo> call, Response<Repo> response) {
                selectedRepo.setValue(response.body());
            }

            @Override
            public void onFailure(Call<Repo> call, Throwable t) {
                Log.e(getClass().getSimpleName(), "Error loading repo", t);
            }
        });
    }
    //</editor-fold>


    @Override
    protected void onCleared() {
        super.onCleared();
        if (repoCall != null) {
            repoCall.cancel();
            repoCall = null;
        }
    }
}

File: app/src/main/java/technologies/acviewmodel/screens/repos/viewmodel/SelectedRepoViewModel.kt
package technologies.acviewmodel.screens.repos.viewmodel

import android.arch.lifecycle.LiveData
import android.arch.lifecycle.MutableLiveData
import android.arch.lifecycle.ViewModel
import android.os.Bundle
import android.util.Log

import javax.inject.Inject

import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import technologies.acviewmodel.model.Repo
import technologies.acviewmodel.networking.RepoService

class SelectedRepoViewModel @Inject
constructor(private val repoService: RepoService) : ViewModel() {

    private val selectedRepo = MutableLiveData<Repo>()
    private var repoCall: Call<Repo>? = null

    fun getSelectedRepo(): LiveData<Repo> {
        return selectedRepo
    }

    fun setSelectedRepo(repo: Repo) {
        selectedRepo.value = repo
    }

    //<editor-fold desc="For saving and restoring data when android kills the process on out of memory exception or process death">
    fun restoreFromBundle(savedInstanceState: Bundle?) {
        if (selectedRepo.value == null) {
            //We only care about restoring if we don't have a selected Repo set already.
            if (savedInstanceState != null && savedInstanceState.containsKey("repo_details")) {
                loadRepo(savedInstanceState.getStringArray("repo_details")!!)
            }
        }
    }

    fun saveToBundle(outState: Bundle) {
        if (selectedRepo.value != null) {
            outState.putStringArray("repo_details", arrayOf(selectedRepo.value!!.owner!!.login, selectedRepo.value!!.name))
        }
    }

    private fun loadRepo(repoDetails: Array<String>) {
        repoCall = repoService.getRepo(repoDetails[0], repoDetails[1])
        repoCall!!.enqueue(object : Callback<Repo> {
            override fun onResponse(call: Call<Repo>, response: Response<Repo>) {
                selectedRepo.value = response.body()
            }

            override fun onFailure(call: Call<Repo>, t: Throwable) {
                Log.e(javaClass.simpleName, "Error loading repo", t)
            }
        })
    }
    //</editor-fold>


    override fun onCleared() {
        super.onCleared()
        if (repoCall != null) {
            repoCall!!.cancel()
            repoCall = null
        }
    }
}
